### PR TITLE
Improvement rating user skill

### LINKED ISSUE #285: Improvement update user skill level
- improve update skill level lẻ 0.5 star, vd 2.5, 3.5 star

image

### PR DESCRIPTION
**Pull request by khaletuanncc**
https://github.com/ncc-erp/ncc-erp-project/issues/285

## Summary by CodeRabbit

- **New Features**
	- Enhanced skill rank display logic across various components for better visual representation.
	- Improved user experience with refined styling and interaction elements in resource management interfaces.
  
- **Bug Fixes**
	- Adjusted logic for skill rank color coding and display conditions to ensure clarity in user skill representation.

- **Documentation**
	- Updated CSS styles for skill rating components to improve layout and appearance.

- **Chores**
	- Streamlined methods related to user skill updates and validations for better performance and maintainability.

### CODE REVIEW

File: aspnet-core/src/ProjectManagement.Application/Users/Dto/UpdateUserSkillDto.cs [hunk 1]
```diff
@@ -14,6 +14,6 @@ public class UpdateUserSkillDto
     public struct SkillPoint
     {
         public long SkillId { get; set; }
-        public SkillRank SkillRank { get; set; }
+        public float SkillRank { get; set; }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Application/Users/Dto/UpdateUserSkillDto.cs:17]: resolved in controller

File: aspnet-core/src/ProjectManagement.Application/APIs/ProjectUserBills/Dto/BillUserSkillDto.cs [hunk 2]
```diff
@@ -14,6 +14,6 @@ public class UpdateUserBillSkillDto
     public struct SkillPoint
     {
         public long SkillId { get; set; }
-        public SkillRank SkillRank { get; set; }
+        public float SkillRank { get; set; }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Application/APIs/ProjectUserBills/Dto/BillUserSkillDto.cs:17]: resolved in controller

File: aspnet-core/src/ProjectManagement.Core/Services/ResourceService/Dto/UserSkillDto.cs [hunk 3]
```diff
@@ -11,7 +11,7 @@ public class UserSkillDto
         public long UserId { get; set; }
         public long SkillId { get; set; }
         public string SkillName { get; set; }
-        public SkillRank SkillRank { get; set; }
+        public float SkillRank { get; set; }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Services/ResourceService/Dto/UserSkillDto.cs:14]: resolved in controller

File: aspnet-core/src/ProjectManagement.Core/Entities/UserSkill.cs [hunk 4]
```diff
@@ -15,7 +15,7 @@ public class UserSkill : FullAuditedEntity<long>, IMayHaveTenant
         public long UserId { get; set; }
         [ForeignKey(nameof(UserId))]
         public User User { get; set; }
-        public SkillRank SkillRank { get; set; }
+        public float SkillRank { get; set; }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Core/Entities/UserSkill.cs:18]: resolved in controller

File: angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.css [hunk 5]
```diff
@@ -16,3 +16,56 @@
...
+}
+
+.rating > label::before {
+    content: "\f005";
+    display: inline-block;
+    font-family: 'Font Awesome 5 Free';
+    font-size: 1.5em;
+    margin-right: .5em;
+    clip-path:inset(0 0 0 50%);
+}
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.css:39]: unnecessary

File: angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.html [hunk 6]
```diff
@@ -42,14 +42,26 @@ <h4>{{!isNotUpdate ? 'Update skill for: ' : 'Skill for: ' }}<strong>{{data.fullN
...
+                                [(ngModel)]="item.skillRank"/>
+                            <label [for]="'star' + (rating.toString().replace('.', '_')) + '_' + i"
+                                [title]="rating + ' stars'"
+                                [ngClass]="{
+                                'half': rating % 1 !== 0,
+                                'reset-rating': item.skillRank === 0
+                                }">
+                            </label>
+                        </ng-container>
+                    </fieldset>
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.html:64]: unnecessary

File: angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.ts [hunk 7]
```diff
@@ -135,20 +132,17 @@ export class UpdateUserSkillDialogComponent implements OnInit {
...
+    if ((item.skillRank === 0.5 && rating === 0.5) || (item.skillRank === 1 && rating === 1)) {
+      item.skillRank = 0;
+      this.snackBar.open('Rating cleared!', '', { duration: this.snackBarDuration });
+      return;
+    }
+    if (item.skillRank === rating) {
+      return;
     }
+    item.skillRank = rating;
+    this.snackBar.open('You rated ' + rating + ' / 5', '', { duration: this.snackBarDuration });
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.ts:145]: unnecessary

File: angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.ts [hunk 8]
```diff
@@ -135,20 +132,17 @@ export class UpdateUserSkillDialogComponent implements OnInit {
...
-    if (item.skillRank>= index + 1) {
-      return 'fa fa-star';
-    } else {
-      return 'far fa-star';
+  onChangeRating(rating: number, item: any) {
+    if ((item.skillRank === 0.5 && rating === 0.5) || (item.skillRank === 1 && rating === 1)) {
+      item.skillRank = 0;
+      this.snackBar.open('Rating cleared!', '', { duration: this.snackBarDuration });
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/users/update-user-skill-dialog/update-user-skill-dialog.component.ts:140]: wrong logic

File: angular/src/app/modules/delivery-management/delivery/available-resource-tab/vendor/vendor.component.html [hunk 9]
```diff
@@ -225,9 +225,14 @@
...
-                                                    yellow:skill.skillRank == 3 ,orange:skill.skillRank == 4}' style="
-                                                    font-size: 13px;"></i> </strong>
+                                                    [ngClass]="{
+                                                        grey: skill.skillRank >= 0 && skill.skillRank <= 2,
+                                                        yellow: skill.skillRank > 2 && skill.skillRank <= 3,
+                                                        orange: skill.skillRank > 3 && skill.skillRank <= 4,
+                                                        crimson: skill.skillRank > 4 && skill.skillRank <= 5
+                                                    }"
+                                                    style="font-size: 16px;"></i>
+                                                </strong>
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/delivery-management/delivery/available-resource-tab/vendor/vendor.component.html:235]: unnecessary

File: angular/src/app/modules/delivery-management/delivery/available-resource-tab/plan-resource/plan-resource.component.html [hunk 10]
```diff
@@ -272,9 +272,14 @@
...
-                                                      yellow:skill.skillRank == 3 ,orange:skill.skillRank == 4}' style="
-                                                      font-size: 13px;"></i> </strong>
+                                                    [ngClass]="{
+                                                        grey: skill.skillRank >= 0 && skill.skillRank <= 2,
+                                                        yellow: skill.skillRank > 2 && skill.skillRank <= 3,
+                                                        orange: skill.skillRank > 3 && skill.skillRank <= 4,
+                                                        crimson: skill.skillRank > 4 && skill.skillRank <= 5
+                                                    }"
+                                                    style="font-size: 16px;"></i>
+                                                </strong>
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/delivery-management/delivery/available-resource-tab/plan-resource/plan-resource.component.html:282]: unnecessary

File: aspnet-core/src/ProjectManagement.Application/APIs/ProjectUserBills/ProjectUserBillAppService.cs [hunk 11]
```diff
@@ -795,7 +795,7 @@ public async Task UpdateBillUserSkill(UpdateUserBillSkillDto input)
             if (user == default)
                 throw new UserFriendlyException($"Can not found project user bill with Id = {input.Id}");
             // check exception
-            if (input.UserSkills.Any(i => i.SkillRank < SkillRank.None || i.SkillRank > SkillRank.Expert))
+            if (input.UserSkills.Any(i => i.SkillRank < 0 || i.SkillRank > 5))
                 throw new UserFriendlyException("Skill rank must be from 1 to 5!");
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Application/APIs/ProjectUserBills/ProjectUserBillAppService.cs:799]: existed value 0.5

File: angular/src/app/modules/delivery-management/delivery/available-resource-tab/all-resource/all-resource.component.html [hunk 12]
```diff
@@ -437,9 +437,14 @@
...
-                                                    yellow:skill.skillRank == 3 ,orange:skill.skillRank == 4}' style="
-                                                    font-size: 13px;"></i></strong>
+                                                    [ngClass]="{
+                                                        grey: skill.skillRank >= 0 && skill.skillRank <= 2,
+                                                        yellow: skill.skillRank > 2 && skill.skillRank <= 3,
+                                                        orange: skill.skillRank > 3 && skill.skillRank <= 4,
+                                                        crimson: skill.skillRank > 4 && skill.skillRank <= 5
+                                                    }"
+                                                    style=" font-size: 16px;"></i>
+                                                </strong>
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/delivery-management/delivery/available-resource-tab/all-resource/all-resource.component.html:447]: unnecessary

File: aspnet-core/src/ProjectManagement.Application/Users/UserAppService.cs [hunk 13]
```diff
@@ -191,7 +191,7 @@ public async Task updateUserSkill(UpdateUserSkillDto input)
             if (user == default)
                 throw new UserFriendlyException($"Can not found user with Id = {input.Id}");
             // check exception
-            if (input.UserSkills.Any(i => i.SkillRank < SkillRank.None || i.SkillRank > SkillRank.Expert))
+            if (input.UserSkills.Any(i => i.SkillRank < 0 || i.SkillRank > 5))
                 throw new UserFriendlyException("Skill rank must be from 1 to 5!");
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply aspnet-core/src/ProjectManagement.Application/Users/UserAppService.cs:195]: existed value 0.5

File: angular/src/app/modules/pm-management/training-projects/training-project-detail/training-resource-management/training-resource-management.component.html [hunk 14]
```diff
@@ -622,13 +621,12 @@
...
                                 [ngClass]="{
-                                  crimson: item.skillRank == 5,
-                                  grey:
-                                    item.skillRank == 1 || item.skillRank == 2,
-                                  yellow: item.skillRank == 3,
-                                  orange: item.skillRank == 4
+                                  grey: item.skillRank >= 0 && item.skillRank <= 2,
+                                  yellow: item.skillRank > 2 && item.skillRank <= 3,
+                                  orange: item.skillRank > 3 && item.skillRank <= 4,
+                                  crimson: item.skillRank > 4 && item.skillRank <= 5
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/pm-management/training-projects/training-project-detail/training-resource-management/training-resource-management.component.html:627]: wrong logic

File: angular/src/app/modules/pm-management/training-projects/training-project-detail/training-resource-management/training-resource-management.component.html [hunk 15]
```diff
@@ -202,13 +202,12 @@
...
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/pm-management/training-projects/training-project-detail/training-resource-management/training-resource-management.component.html:208]: wrong logic

File: angular/src/app/modules/pm-management/list-project/list-project-detail/resource-management/resource-management.component.html [hunk 16]
```diff
@@ -241,13 +241,12 @@
...
-                                    item.skillRank == 1 || item.skillRank == 2,
-                                  yellow: item.skillRank == 3,
-                                  orange: item.skillRank == 4
+                                  grey: item.skillRank >= 0 && item.skillRank <= 2,
+                                  yellow: item.skillRank > 2 && item.skillRank <= 3,
+                                  orange: item.skillRank > 3 && item.skillRank <= 4,
+                                  crimson: item.skillRank > 4 && item.skillRank <= 5
                                 }"
-                                style="font-size: 13px"
+                                style="font-size: 16px"
```
[the remark this answers is not in the document]
Comment (khaletuanncc) [reply angular/src/app/modules/pm-management/list-project/list-project-detail/resource-management/resource-management.component.html:249]: wrong logic

**Review by tiennguyenhuuncc — APPROVED** [review]

**Pull request merged**